### PR TITLE
Support GARD OD metrics

### LINKED ISSUE #132: Support GARD Amory Object Detection Metrics
Add support to use https://dbolya.github.io/tide/ for OD metrics.

### PR DESCRIPTION
**Pull request by treubig26**
Closes #132 (Add support to use https://dbolya.github.io/tide/ for OD metrics.)

### CODE REVIEW

File: library/src/armory/metrics/detection.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            )
+            misclassification_rate = (
+                misclassification_array.mean()
+                if len(misclassification_array) > 0
+                else torch.tensor(0)
+            )
+
+            # Any ground-truth box that had no overlapping predicted box is considered a
+            # disappearance
+            disappearance_rate = 1 - true_positive_rate - misclassification_rate
```
Comment (swsuggs) [inline library/src/armory/metrics/detection.py:157]: I realize this is copied from Armory, but I think we should define disappearance_rate to be 0 if there are no ground truth boxes (```len(targets)==0```).  Currently, disappearance rate is 1 if there are no targets, which doesn't make sense to me.  This would also require updating the unit test for the example with the empty target.

File: library/src/armory/metrics/tide.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        data_preds.max_dets = max_dets
+        data_target.max_dets = max_dets
+
+        tide = tidecv.TIDE()
+        tide.evaluate_range(data_target, data_preds, mode=tidecv.TIDE.BOX)
+
+        errors = tide.get_all_errors()
+
+        return {
+            "mAP": {x.pos_thresh: x.ap / 100 for x in tide.run_thresholds["preds"]},
```
Comment (mwartell) [inline library/src/armory/metrics/tide.py:82]: Are these dict keys consumed by something else, or would it be better spelled out as `mean_average_precision` etc.?

Comment (mwartell) [reply library/src/armory/metrics/tide.py:82]: also `FalseNeg`

File: library/src/armory/metrics/detection.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+            # Initialize arrays that will indicate whether each respective ground-truth
+            # box is a true positive or misclassified
+            true_positive_array = torch.zeros((num_gt_boxes,))
+            misclassification_array = torch.zeros((num_gt_boxes,))
+
+            # Only consider the model's confident predictions
+            assert y_pred["scores"] is not None
+            conf_pred_indices = torch.where(y_pred["scores"] > self.score_threshold)[0]
+            if self.class_list:
```
Comment (deprit) [inline library/src/armory/metrics/detection.py:106]: This test fails if the class list was specified as `class_list=(0,)` in the constructor. Since object detectors assign class 0 to background by convention, this edge case might not make sense.

File: library/tests/unit/test_metrics_detection.py
Comment (deprit) [inline library/tests/unit/test_metrics_detection.py:1]: The current test cases cover the `mean` constructor parameter but ignore the parameters below:
- iou_threshold,
- score_threshold,
- class_list.

**Review by deprit — CHANGES_REQUESTED** [review]

File: library/src/armory/metrics/tide.py (hunk 2 above)
Comment (treubig26) [reply library/src/armory/metrics/tide.py:82]: I used the same keys we had in GARD-armory. They are not directly consumed by anything (they would be if you had any post-evaluation analysis). As a point of comparison, torchmetrics uses the acronyms "map" and "mar" in their MeanAveragePrecision metric.

File: library/src/armory/metrics/detection.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        iou_threshold: float = 0.5,
+        score_threshold: float = 0.5,
+        class_list: Optional[Sequence[int]] = None,
+    ):
+        """
+        Initialize the metric.
+
+        Args:
+            mean: `True` to return the mean values across all images or `False`
+                to return the individual values per image
```
[the remark this answers is not in the document]
Comment (treubig26) [reply library/src/armory/metrics/detection.py:50]: MLFlow has a limitation in that _all_ metrics have to be scalar values. You _could_ treat the array indices as step values for the MLFlow metric--but that's just abusing the API I think (it has been recognized as a potential workaround, but not applicable in all cases: https://github.com/mlflow/mlflow/issues/2451).

We had a discussion about this many months ago during JATIC and we had said then that we were ok with metrics being wonky as presented in MLFlow because we didn't expect anyone to be trying to use them within MLFlow itself. If that assumption is being changed, then we certainly can revisit that (which has implications on all metrics from torchmetrics, not just this one). It strikes me as odd (i.e., variable behavior) if some metrics are directly recorded in MLFlow (i.e., when scalar) but others are simply attached as artifacts (i.e., when dicts & lists), but that's not a showstopper.

**Review by deprit — APPROVED** [review]
